File: 8. Multiple Pattern Search (Aho–Corasick Algorithm).cpp
// Implement the Aho–Corasick algorithm to search multiple patterns simultaneously in a given text.

#include <bits/stdc++.h>
using namespace std;

struct TrieNode
{
    unordered_map<char, TrieNode*> children;
    TrieNode* failure = nullptr;
    vector<int> output;  // store indexes of patterns ending here
};

class AhoCorasick
{
   private:
    TrieNode* root;

   public:
    AhoCorasick() { root = new TrieNode(); }

    void insert(const string& pattern, int index)
    {
        TrieNode* node = root;
        for (char c : pattern)
        {
            if (!node->children.count(c))
                node->children[c] = new TrieNode();
            node = node->children[c];
        }
        node->output.push_back(index);
    }

    void buildFailureLinks()
    {
        queue<TrieNode*> q;
        root->failure = root;
        for (auto& p : root->children)
        {
            p.second->failure = root;
            q.push(p.second);
        }

        while (!q.empty())
        {
            TrieNode* curr = q.front();
            q.pop();
            for (auto& p : curr->children)
            {
                char c = p.first;
                TrieNode* child = p.second;

                TrieNode* fail = curr->failure;
                while (fail != root && !fail->children.count(c)) fail = fail->failure;
                if (fail->children.count(c))
                    fail = fail->children[c];

                child->failure = fail;

                // merge output
                child->output.insert(child->output.end(), fail->output.begin(), fail->output.end());

                q.push(child);
            }
        }
    }

    void search(const string& text, vector<string>& patterns)
    {
        TrieNode* node = root;
        for (int i = 0; i < text.size(); i++)
        {
            char c = text[i];
            while (node != root && !node->children.count(c)) node = node->failure;

            if (node->children.count(c))
                node = node->children[c];

            for (int idx : node->output)
                cout << "Pattern \"" << patterns[idx] << "\" found at index "
                     << i - patterns[idx].size() + 1 << endl;
        }
    }
};

int main()
{
    vector<string> patterns = {"he", "she", "his", "hers"};
    string text = "ushers";

    AhoCorasick ac;
    for (int i = 0; i < patterns.size(); i++) ac.insert(patterns[i], i);

    ac.buildFailureLinks();
    ac.search(text, patterns);

    return 0;
}

/*
Time Complexity (TC): O(N + Σ|Pi| + K)
- N = length of text
- Σ|Pi| = total length of all patterns
- K = total number of matches found
Space Complexity (SC): O(Σ|Pi| * alphabet_size) due to Trie nodes
*/

File: 8. Multiple Pattern Search (Aho–Corasick Algorithm).py
# Implement the Aho–Corasick algorithm to search multiple patterns simultaneously in a given text.
from collections import deque, defaultdict

class TrieNode:
    def __init__(self):
        self.children = {}
        self.failure = None
        self.output = []

class AhoCorasick:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, pattern, index):
        node = self.root
        for c in pattern:
            if c not in node.children:
                node.children[c] = TrieNode()
            node = node.children[c]
        node.output.append(index)

    def build_failure_links(self):
        queue = deque()
        self.root.failure = self.root
        for child in self.root.children.values():
            child.failure = self.root
            queue.append(child)

        while queue:
            node = queue.popleft()
            for c, child in node.children.items():
                fail = node.failure
                while fail != self.root and c not in fail.children:
                    fail = fail.failure
                if c in fail.children:
                    fail = fail.children[c]
                child.failure = fail
                child.output += fail.output
                queue.append(child)

    def search(self, text, patterns):
        node = self.root
        for i, c in enumerate(text):
            while node != self.root and c not in node.children:
                node = node.failure
            if c in node.children:
                node = node.children[c]
            for idx in node.output:
                print(f'Pattern "{patterns[idx]}" found at index {i - len(patterns[idx]) + 1}')

# Example usage
patterns = ["he", "she", "his", "hers"]
text = "ushers"

ac = AhoCorasick()
for i, p in enumerate(patterns):
    ac.insert(p, i)
ac.build_failure_links()
ac.search(text, patterns)

"""
Time Complexity (TC): O(N + Σ|Pi| + K)
- N = length of text
- Σ|Pi| = total length of all patterns
- K = total number of matches found
Space Complexity (SC): O(Σ|Pi| * alphabet_size) due to Trie nodes
"""
